examples/capture/main.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2023 The Ebitengine Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"image/color"
+	"log"
+	"math"
+	"runtime"
+
+	"github.com/hajimehoshi/ebiten/v2"
+	"github.com/hajimehoshi/ebiten/v2/ebitenutil"
+	"github.com/hajimehoshi/ebiten/v2/inpututil"
+)
+
+var pointerImage = ebiten.NewImage(8, 8)
+
+func init() {
+	pointerImage.Fill(color.RGBA{0xff, 0, 0, 0xff})
+}
+
+const (
+	screenWidth  = 640
+	screenHeight = 480
+)
+
+type Game struct {
+	mouseX int
+	mouseY int
+	x      int
+	y      int
+}
+
+func (g *Game) Update() error {
+	if ebiten.CursorMode() == ebiten.CursorModeCaptured {
+		cursorX, cursorY := ebiten.CursorPosition()
+
+		if g.mouseX == math.MinInt32 && g.mouseY == math.MinInt32 {
+			// Initialize first position to establish delta.
+			if cursorX != 0 && cursorY != 0 {
+				g.mouseX, g.mouseY = cursorX, cursorY
+			}
+		} else {
+			deltaX, deltaY := cursorX-g.mouseX, cursorY-g.mouseY
+			g.mouseX, g.mouseY = cursorX, cursorY
+
+			if deltaX != 0 {
+				g.x += deltaX
+			}
+
+			if deltaY != 0 {
+				g.y += deltaY
+			}
+		}
+
+		// Constrain red dot within screen view.
+		if g.x < 0 {
+			g.x = 0
+		} else if g.x > screenWidth-pointerImage.Bounds().Dx() {
+			g.x = screenWidth - pointerImage.Bounds().Dx()
+		}
+
+		if g.y < 0 {
+			g.y = 0
+		} else if g.y > screenHeight-pointerImage.Bounds().Dy() {
+			g.y = screenHeight - pointerImage.Bounds().Dy()
+		}
+	}
+
+	if inpututil.IsKeyJustPressed(ebiten.KeySpace) {
+		if ebiten.CursorMode() == ebiten.CursorModeCaptured {
+			// Release mouse cursor capture.
+			ebiten.SetCursorMode(ebiten.CursorModeVisible)
+		} else {
+			// Recapture mouse cursor.
+			ebiten.SetCursorMode(ebiten.CursorModeCaptured)
+
+			// Reset mouse cursor position for its return.
+			g.mouseX, g.mouseY = math.MinInt32, math.MinInt32
+		}
```

Comment:
Let's refactor this logic out to a new function like `updateCursorMode`

```go
func (g *Game) updateCursorMode(cursorMode ebiten.CursorModeType) {
	if ebiten.CursorMode() == cursorMode {
		return
	}
	ebiten.SetCursorMode(mode)
	if cursorMode == ebiten.CursorModeCaptured {
		// Initialize first position to establish delta.
		g.mouseX, g.mouseY = ebiten.CursorPosition()
	}
}
```

```go
	if inpututil.IsKeyJustPressed(ebiten.KeySpace) {
		if ebiten.CursorMode() == ebiten.CursorModeVisible {
			g.updateCursorMode(ebiten.CursorModeCaptured)
		} else {
			g.updateCursorMode(ebiten.CursorModeVisible)
		}
	}
```

Also, let's call this from `main`

```go
	if runtime.GOOS != "js" {
		g.updateCursorMode(ebiten.CursorModeCaptured)
	}
```

Then, we don't have to have `MinInt32` as a special value.

Reply:
I like the idea of the function, the only problem is that by not initializing the special value `MinInt32` in `main` the cursor start position ends up being thrown to the bottom right corner `632, 472` instead of the center as desired `(screenWidth/2, screenHeight/2)`.

Reply:
Actually it seems to appear the the start position of the red dot depends on the position where the mouse happens to be on the desktop when the program first starts up. My preference would be the always have the dot start in the center, however I'll defer to you. If the start position doesn't bother you I'll go ahead with your idea.

Reply:
Oh right, I realized `ebiten.CursorPosition` is not valid before `ebiten.RunGame`.

What about using `sync.Once`?:

```go
type Game struct {
	// ...
	initOnce sync.Once
}

func (g*Game) Update() error {
	g.initOnce.Do(func() {
		g.mouseX, g.mouseY = ebiten.CursorPosition()
	})
	// ...
	if inpututil.IsKeyJustPressed(ebiten.KeySpace) {
		if ebiten.CursorMode() == ebiten.CursorModeVisible {
			ebiten.SetCursorMode(ebiten.CursorModeCaptured)
		} else {
			ebiten.SetCursorMode(ebiten.CursorModeVisible)
			g.mouseX, g.mouseY = ebiten.CursorPosition()
		}
	}
}
```

`main` remains not to use `updateCursorMode`:

```go
	if runtime.GOOS != "js" {
		ebiten.SetCursorMode(ebiten.CursorModeCaptured)
	}
```